pkg/controllers/networkpolicy/policy_combiner.go

```diff
@@ -0,0 +1,61 @@
+package networkpolicy
+
+import (
+	log "github.com/sirupsen/logrus"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sort"
+)
+
+func combinePolicies(nps []*networkingv1.NetworkPolicy) *networkingv1.NetworkPolicy {
+	// policies should be combined in alphabetical order
+
+	sort.Slice(nps, func(i, j int) bool {
+		return nps[i].Name < nps[j].Name
+	})
+
+	base := &networkingv1.NetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nps[0].Labels[dynamicLabel],
+			Namespace: nps[0].Namespace,
+		},
+		Spec: networkingv1.NetworkPolicySpec{
+			PodSelector: nps[0].Spec.PodSelector,
+			PolicyTypes: nps[0].Spec.PolicyTypes,
+		},
+	}
+
+	baseSelector := base.Spec.PodSelector.String()
+	for _, np := range nps {
+		if np.Namespace != base.Namespace {
+			log.Warn("Namespaces aren't equal for all network policies labelled with monzo.com/dynamic-networkpolicy=", base.Name, ". skipping ", np.Name)
+			continue
+		}
+
+		if np.Spec.PodSelector.String() != baseSelector {
+			log.Warn("Selectors aren't equal for all network policies labelled with monzo.com/dynamic-networkpolicy=", base.Name, ". skipping ", np.Name)
+			continue
+		}
+
+		base.Spec.Ingress = append(base.Spec.Ingress, np.Spec.Ingress...)
+		base.Spec.Egress = append(base.Spec.Egress, np.Spec.Egress...)
+	}
+
+	return base
+}
+
+func interfacesToNetPols(is []interface{}) []*networkingv1.NetworkPolicy {
+	nps := make([]*networkingv1.NetworkPolicy, 0, len(is))
+	for _, i := range is {
+		nps = append(nps, i.(*networkingv1.NetworkPolicy))
+	}
+	return nps
+}
+
+func combinerIndexFunc(obj interface{}) ([]string, error) {
```

Comment:
So the docs for the indexing function (https://godoc.org/k8s.io/client-go/tools/cache#IndexFunc) say
>IndexFunc knows how to compute the set of indexed values for an object.

So does this mean that for this key (the NetworkPolicy object) we are indexing the value of `dynamicLabel`. Which I guess will be the name of the "group" of policies? This doesn't make sense to me.

Reply:
My thinking is we have something like `monzo.com/dynamic-networkpolicy: ingress.s-ledger` which will produce a calico network policy called `knp.default.ingress.s-ledger`, which is produced by combining all network policies with that label. Unlabelled policies are not affected.

Reply:
As we discussed, what we're returning here are actually the keys by which we want to index the given obj. It makes sense to me now.